src/ResourceManagement/ResourceManager/GenericResource/GenericResourcesImpl.cs

```diff
@@ -293,5 +271,29 @@ public async Task<IPagedCollection<IGenericResource>> ListByTagAsync(string reso
                 Manager.Inner.Resources.ListByResourceGroupNextAsync,
                 WrapModel, loadAllPages, cancellationToken);
         }
+
+        internal static string GetApiVersion(string id, IResourceManager resourceManager)
+        {
+            string apiVersionValue;
+            string resourceProvider = ResourceUtils.ResourceProviderFromResourceId(id);
+            string resourceType = ResourceUtils.ResourceTypeForApiVersion(id);
+            string apiVersionKey = resourceProvider + "/" + resourceType;
+            if (cachedApiVersions.ContainsKey(apiVersionKey))
+            {
+                apiVersionValue = cachedApiVersions[apiVersionKey];
+            }
+            else
+            {
+                apiVersionValue = ResourceUtils.DefaultApiVersionFromResourceId(
+                    id,
+                    resourceManager.Providers.GetByName(resourceProvider));
+                if (apiVersionValue == null)
+                {
+                    throw new ArgumentException(string.Format("{0} is not a supported resource type in provider {1}", resourceType, resourceProvider));
+                }
+                cachedApiVersions.Add(apiVersionKey, apiVersionValue);
```

Comment:
DO NOT call Add in concurrency (it throws exception when find key exists). Use TryAdd or AddOrUpdate.